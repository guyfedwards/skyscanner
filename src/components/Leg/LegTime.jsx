import React from 'react';
import BpkText from 'bpk-component-text';
import styled from 'styled-components';
import propTypes from 'prop-types';

const Root = styled.div`
  display: flex;
  flex-direction: column;
  width: 40px;

  margin-right: 15px;
  margin-left: 15px;
`;

const Airport = styled(BpkText)`
  color: #aba9a9;
`;

const LegTime = ({ time, airport }) => {
  const localTime = new Date(time).toLocaleTimeString('en-GB', {
    hour: '2-digit',
    minute: '2-digit',
  });

  return (
    <Root>
      <BpkText textStyle="lg">{localTime}</BpkText>
      <Airport textStyle="lg">{airport}</Airport>
    </Root>
  );
};

LegTime.propTypes = {
  airport: propTypes.string.isRequired,
  time: propTypes.string.isRequired,
};

export default LegTime;
