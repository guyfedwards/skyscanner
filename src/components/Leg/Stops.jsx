import React from 'react';
import propTypes from 'prop-types';
import styled from 'styled-components';

const Root = styled.span`
  color: ${({ direct }) => (direct ? '#00a698' : '#d1435b')};
`;

const Stops = ({ stops }) => {
  const direct = stops === 0;
  const text = direct ? 'Direct' : `${stops} stop ${stops > 1 ? 's' : ''}`;

  return <Root direct={direct}>{text}</Root>;
};

Stops.propTypes = {
  stops: propTypes.number.isRequired,
};

export default Stops;
