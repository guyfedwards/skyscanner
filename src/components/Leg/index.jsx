import React from 'react';
import styled from 'styled-components';
import BpkImage from 'bpk-component-image';
import BpkSmallArrow from 'bpk-component-icon/sm/long-arrow-right';
import BpkText from 'bpk-component-text';

import legType from '../../types/leg';

import LegTime from './LegTime';
import Stops from './Stops';

const Root = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;

  & + & {
    margin-top: 10px;
  }
`;

const Left = styled.div`
  display: flex;
`;

const StyledImage = styled(BpkImage)`
  height: 32px;
  width: 32px;
`;

const LegTimes = styled.div`
  display: flex;
  flex-wrap: nowrap;
  align-items: center;
`;

const Right = styled.div`
  display: flex;
  flex-direction: column;
`;

const Arrow = styled.div`
  margin-right: 5px;
  margin-left: 5px;
`;

const StyledBpkSmallArrow = styled(BpkSmallArrow)`
  fill: #aba9a9;
`;

const Duration = styled(BpkText)`
  color: #aba9a9;
`;

const Leg = ({ leg }) => {
  const durationHours = Math.floor(leg.duration_mins / 60);
  const durationMins = leg.duration_mins % 60;

  return (
    <Root>
      <Left>
        <StyledImage
          altText={leg.airline_name}
          height={32}
          width={32}
          src={`https://logos.skyscnr.com/images/airlines/favicon/${leg.airline_id}.png`}
        />
        <LegTimes>
          <LegTime time={leg.departure_time} airport={leg.departure_airport} />
          <Arrow>
            <StyledBpkSmallArrow />
          </Arrow>
          <LegTime time={leg.departure_time} airport={leg.departure_airport} />
        </LegTimes>
      </Left>
      <Right>
        <Duration>
          {durationHours}h {durationMins}
        </Duration>
        <Stops stops={leg.stops} />
      </Right>
    </Root>
  );
};

Leg.propTypes = legType.isRequired;

export default Leg;
