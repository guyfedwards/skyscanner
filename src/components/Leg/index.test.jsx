import React from 'react';
import ReactDOM from 'react-dom';
import ShallowRenderer from 'react-test-renderer/shallow';

import Leg from '.';

const leg = {
  id: 'leg_1',
  departure_airport: 'BUD',
  arrival_airport: 'LTN',
  departure_time: '2020-10-31T15:35',
  arrival_time: '2020-10-31T17:00',
  stops: 0,
  airline_name: 'Wizz Air',
  airline_id: 'WZ',
  duration_mins: 145,
};

describe('Leg', () => {
  it('should render without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(<Leg leg={leg} />, div);
  });

  it('should render correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<Leg leg={leg} />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
