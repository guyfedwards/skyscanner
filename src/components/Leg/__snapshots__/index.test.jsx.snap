// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leg should render correctly 1`] = `
<ForwardRef(styled.div)>
  <ForwardRef(styled.div)>
    <ForwardRef(Styled(BpkImage))
      altText="Wizz Air"
      height={32}
      src="https://logos.skyscnr.com/images/airlines/favicon/WZ.png"
      width={32}
    />
    <ForwardRef(styled.div)>
      <LegTime
        airport="BUD"
        time="2020-10-31T15:35"
      />
      <ForwardRef(styled.div)>
        <ForwardRef(Styled(_default)) />
      </ForwardRef(styled.div)>
      <LegTime
        airport="BUD"
        time="2020-10-31T15:35"
      />
    </ForwardRef(styled.div)>
  </ForwardRef(styled.div)>
  <ForwardRef(styled.div)>
    <ForwardRef(Styled(BpkText))>
      2
      h 
      25
    </ForwardRef(Styled(BpkText))>
    <Stops
      stops={0}
    />
  </ForwardRef(styled.div)>
</ForwardRef(styled.div)>
`;
